lib/core/view_models/login_view_model.dart

```diff
@@ -1,54 +1,54 @@
-import 'package:flutter/material.dart';
 import 'package:pedantic/pedantic.dart';
 import 'package:provider_start/core/constant/view_routes.dart';
 import 'package:provider_start/core/enums/view_state.dart';
 import 'package:provider_start/core/exceptions/auth_exception.dart';
 import 'package:provider_start/core/mixins/validators.dart';
 import 'package:provider_start/core/services/auth/auth_service.dart';
 import 'package:provider_start/core/services/navigation/navigation_service.dart';
+import 'package:provider_start/core/utils/simple_log_printer.dart';
 import 'package:provider_start/core/view_models/base_view_model.dart';
 import 'package:provider_start/locator.dart';
 
 class LoginViewModel extends BaseViewModel with Validators {
-  final _navigationService = locator<NavigationService>();
+  final _log = getLogger();
   final _authService = locator<AuthService>();
+  final _navigationService = locator<NavigationService>();
 
-  final _emailController = TextEditingController();
-  final _passwordController = TextEditingController();
-  final _passwordFocusNode = FocusNode();
-  final _formKey = GlobalKey<FormState>();
-  final _viewFocusNode = FocusNode();
+  String _email = '';
+  String get email => _email;
 
-  TextEditingController get emailController => _emailController;
-  TextEditingController get passwordController => _passwordController;
-  FocusNode get passwordFocusNode => _passwordFocusNode;
-  GlobalKey<FormState> get formKey => _formKey;
-  FocusNode get viewFocusNode => _viewFocusNode;
+  String _password = '';
+  String get password => _password;
 
-  Future<void> login() async {
-    if (!_formKey.currentState.validate()) return;
+  bool _canSubmitEmail = false;
+  bool _canSubmitPassword = false;
+  bool get canLoginUser => _canSubmitEmail && _canSubmitPassword;
 
-    setState(ViewState.Busy);
+  void setEmail(String email) {
+    _log.d('email:$email');
+    _email = email;
+    _canSubmitEmail = validateEmail(email) == null;
+    notifyListeners();
+  }
+
+  void setPassword(String password) {
+    _log.d('password:$password');
+    _password = password;
+    _canSubmitPassword = validatePassword(password) == null;
+    notifyListeners();
+  }
```

Comment:
So the TextFormField is a stateful widget that updates its own state when the value of password is updated. Our view model doesn't need to notifyListeners each time the password is updated.

Reply:
Right, I left it there by mistake cause I use to manage the state myself with that. I don't like the validation messages of TextFormField, I use custom messages controlled by me.

Reply:
I think we can completely remove these functions and refactor the login function to accept the email and password. https://www.filledstacks.com/post/firebase-authentication-in-flutter/ is probably the example we want to follow.